nemo_skills/mcp/clients.py

```diff
@@ -0,0 +1,574 @@
+                },
+            }
+        ]
+    })
+    manager = build_client_manager(cfg)
+    tools = await manager.list_all_tools()
+    result = await manager.execute_tool("exa_mcp.web_search_exa", {"query": "nemo skills"})
+    ```
+
+    Example (manual usage):
+    ```python
+    client = MCPStreamableHttpClient(base_url="https://mcp.example.com/mcp")
+    tools = await client.list_tools()
+    result = await client.call_tool("some_tool", {"x": 1})
+    ```
+    """
+
+    def __init__(self, base_url: str):
+        self.base_url = base_url
+        self.tools: List[Dict[str, Any]] = []
+
+    async def list_tools(self):
+        async with streamablehttp_client(self.base_url) as (read_stream, write_stream, _):
+            async with ClientSession(read_stream, write_stream) as session:
+                await session.initialize()
+                tools_resp = await session.list_tools()
+                tools_list: List[Dict[str, Any]] = []
+                # tools_resp.tools is expected to be a list of Tool objects
+                for t in getattr(tools_resp, "tools", []) or []:
+                    # Support both input_schema (python) and inputSchema (wire)
+                    input_schema = getattr(t, "input_schema", None)
+                    if input_schema is None:
+                        input_schema = getattr(t, "inputSchema", None)
+                    tools_list.append(
+                        {
+                            "name": getattr(t, "name", None),
+                            "description": getattr(t, "description", ""),
+                            "input_schema": input_schema,
+                        }
+                    )
+                self.tools = tools_list
+                return self.tools
+
+    async def call_tool(self, tool: str, args: dict) -> Any:
+        self._assert_tool_allowed(tool)
+        async with streamablehttp_client(self.base_url) as (read_stream, write_stream, _):
+            async with ClientSession(read_stream, write_stream) as session:
+                await session.initialize()
+                result = await session.call_tool(tool, arguments=args)
+                return struct if (struct := result.structuredContent) is not None else result
+
+
+class MCPStdioClient(MCPClient):
+    """MCP client that launches a server via stdio.
+
+    Args:
+        command: Executable to launch (e.g., "python").
+        args: Command-line arguments (e.g., ["-m", "nemo_skills.mcp.servers.python_tool"]).
+
+    Behavior:
+    - list_tools() fetches tool metadata from the running stdio server.
+    - call_tool() returns the server's structuredContent.
+
+    The following optional configurables can be supplied (injected by the
+    metaclass): hide_args, disabled_tools, enabled_tools, output_formatter,
+    init_hook.
+
+    Example (OmegaConf config, like in mcp_demo.py):
+    ```python
+    from omegaconf import OmegaConf
```

Comment:
do you want to put these examples as tests?

Reply:
Can I depend on the sandbox during pytests on CI? I wrote some that are agnostic to the Sandbox, but if I can use it, then I can test the integration too.